### PR TITLE
Correct the initial rendering of non-scrollable dropdown menus

### LINKED ISSUE #9924: Rendering Dropdown items, first correctly than all blanks and then correctly again
## Steps to Reproduce
There is a visual issue with a dropdown component. When I click the dropdown, the menu expands and it renders the items, then it renders them again but all items are blank; than, it renders them again correctly. 
Sometimes I get an exception. You can see the first lines of the exception below. The full stack trace is in attachment.

You can find a main.dart in attachment to reproduce the issue.

The animated gif comes from a real device (Note 4)

## Logs
I/flutter ( 5406): The following assertion was thrown during performLayout():
I/flutter ( 5406): SliverGeometry has a layoutExtent that exceeds its paintExtent.
I/flutter ( 5406): The paintExtent is 250.3445958314173, but the layoutExtent is 250.34459583141734. Maybe you have
I/flutter ( 5406): fallen prey to floating point rounding errors, and should explicitly apply the min() or max()
I/flutter ( 5406): functions, or the clamp() method, to
dd_issue
 the layoutExtent?

## Flutter Doctor
[√] Flutter (on Microsoft Windows [Version 10.0.14393], channel master)
    • Flutter at C:\Users\Simone\flutter
    • Framework revision ebd79100c0 (83 minutes ago), 2017-05-09 17:30:05 +0200
    • Engine revision cbbf956fcb
    • Tools Dart version 1.23.0-dev.11.11

[√] Android toolchain - develop for Android devices (Android SDK 25.0.2)
    • Android SDK at C:\Users\Simone\AppData\Local\Android\sdk
    • Platform android-25, build-tools 25.0.2
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version: OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[√] Android Studio (version 2.3)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Gradle version 3.2
    • Java version: OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[√] IntelliJ IDEA Community Edition (version 2016.3)
    • Dart plugin version 163.13137
    • Flutter plugin version 12.1

[√] IntelliJ IDEA Ultimate Edition (version 2017.1)
    • Dart plugin version 171.4424
    • Flutter plugin version 13.1

[√] Connected devices
    • Android SDK built for x86 • emulator-5554 • android-x86 • Android 7.0 (API 24) (emulator)
[main_dart.txt](https://github.com/flutter/flutter/files/987372/main_dart.txt)
[full_stack_trace.txt](https://github.com/flutter/flutter/files/987374/full_stack_trace.txt)

### PR DESCRIPTION
**Pull request by HansMuller**
Compute the dropdown menu's location, dimensions and scroll offset when its page is built, rather than at layout time.

Fixes https://github.com/flutter/flutter/issues/9924

### CODE REVIEW

**Comment by HansMuller** [comment]
@Hixie

File: packages/flutter/lib/src/material/dropdown.dart [hunk 1]
```diff
@@ -217,41 +218,19 @@ class _DropdownMenuRouteLayout<T> extends SingleChildLayoutDelegate {
...
-        assert(top >= 0.0);
-        assert(top + childSize.height <= size.height);
+        assert(menuTop >= 0.0);
+        assert(menuTop + menuHeight <= size.height);
       }
       return true;
     });
 
-    if (route.initialLayout) {
-      route.initialLayout = false;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/dropdown.dart]: This code only ran once, right? I really don't understand why touching the menus broke them.

File: packages/flutter/lib/src/material/dropdown.dart [hunk 2]
```diff
@@ -290,14 +269,18 @@ class _DropdownRoute<T> extends PopupRoute<_DropdownRouteResult<T>> {
...
 
-  final ScrollController scrollController = new ScrollController();
   final List<DropdownMenuItem<T>> items;
   final Rect buttonRect;
   final int selectedIndex;
   final int elevation;
   final ThemeData theme;
   final TextStyle style;
 
+  // These fields are used by _DropdownMenuRouteLayout and initialized by buildPage().
```
Comment (Hixie) [inline packages/flutter/lib/src/material/dropdown.dart]: do we guarantee that _DropdownMenuRouteLayout will relayout if buildPage is called again?

File: packages/flutter/lib/src/material/dropdown.dart [hunk 3]
```diff
@@ -196,6 +195,8 @@ class _DropdownMenuRouteLayout<T> extends SingleChildLayoutDelegate {
   final _DropdownRoute<T> route;
 
   Rect get buttonRect => route.buttonRect;
+  double get menuTop => route.menuTop;
+  double get menuHeight => route.menuHeight;
   int get selectedIndex => route.selectedIndex;
   ScrollController get scrollController => route.scrollController;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/dropdown.dart]: It seems like all of these values should really come from the State, which should get them from the Route... getting them like this seems very likely to mean we don't always get laid out again when they change. If we somehow guarantee that we do, we should add a comment here saying how that guarantee is made.

**Comment by Hixie** [comment]
This should have a test to make sure the menu positioning (all the new code) is correct. (It doesn't have to verify that the #9924 bug is fixed, since that will be dealt with when we stop the scrolling thing.)

**Comment by Hixie** [comment]
LGTM modulo above comments.

File: packages/flutter/lib/src/material/dropdown.dart (hunk 1 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/dropdown.dart]: I don't understand why setting the scrollOffset to a non-zero value with a framecallback, or even earlier in buildPage() caused the bug. Hopefully adding support for initially-overscrolled menus (in a separate PR) will shed some light on the root of the problem.

File: packages/flutter/lib/src/material/dropdown.dart (hunk 2 above)
Comment (HansMuller) [inline packages/flutter/lib/src/material/dropdown.dart]: Good point. Although it seems to DTRT, _DropdownMenuRouteLayout.shouldRelayout() really should be comparing all of the values it depends on.

**Pull request merged**